### ISSUE TITLE
Add support for JSON.net JsonSerializerSettings

### ISSUE THREAD

**Issue by j82w** [Azure/azure-cosmos-dotnet-v3#551]
To override a single JsonSerializerSettings requires implementing an entire serializer class. There should be a way to override the default serializer settings to avoid forcing users to create a entire new class just for one setting.

1. Make CosmosSerializerCore public sealed with a constructor to take in the settings.
2. Add an option to CosmosClientOptions. Might be confusing with the ability to override serializer

**Comment by nheinbaugh** [comment]
It would also be nice to have a sample showing a simple change to the default serializer. 

Without official documentation or a sample I don't even know what I don't know. Having to dig through source code and tests is painful

**Comment by j82w** [comment]
@nheinbaugh thanks for the suggestion. I created an issue to track it #559

**Comment by richstokoe** [comment]
Until Microsoft exposes the serializer settings, you could just take the code of the default CosmosJsonSerializer and pass in a JsonSerializerSettings. Example gist: https://gist.github.com/richstokoe/c82fc831c6b4020926f5b5f772f7cd70

**Comment by j82w** [comment]
@richstokoe that is the current plan. That is why this issue was created.

**Comment by richstokoe** [comment]
Sure, and I didn't mean to hinder that, just wanted to show that open source means we don't always have to wait for the samples. Keep going, v3 SDK is awesome.

**Comment by kirankumarkolli** [comment]
JSON.net implementation as public contract will force its dependency even in future when implementation moved to .NET core JSON parser. Sample with JsonSerializer example would help unblock.

**Comment by j82w** [comment]
@richstokoe 3.1.0 is now released with the new serializer

**Comment by j82w** [comment]
@richstokoe and @nheinbaugh can you please take a look at this PR #650? We need to remove the CosmosJsonDotNetSerializer.  The Azure central SDK team has guideline that we need to move to the new system.text.json, and exposing the JSON settings directly conflicts with that. This PR add supports for some common JSON settings. Does this support all of your scenarios?

**Comment by richstokoe** [comment]
@j82w Thanks for the engagement!
I may have missed it in the PR but I’d personally prefer to disambiguate property casing, i.e.
CosmosPropertyNamingPolicy.Default
CosmosPropertyNamingPolicy.UpperCamelCase
CosmosPropertyNamingPolicy.LowerCamelCase
That would solve my original issue (POCO “Id” properties not serialising to “id” without having to add a JsonPropertyAttribute) and would be explicit for consumers.

**Comment by j82w** [comment]
@richstokoe camel case refers to the first character being lower case. Pascal case is where the first character is capitalized. Check out [Camel Case vs Pascal Case](https://www.quora.com/What-is-the-difference-between-Pascal-Case-and-Camel-Case). I don't see an option in Newtonsoft or system.text.JSON that support pascal case by default. They both would require a custom converter. Do you have any requirements to support pascal case?

**Comment by j82w** [comment]
The final solution with cosmos serialization settings it [merged](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/650).

**Comment by adstep** [comment]
Is there support for customer converts with the current setup? I am looking to add a stringtoenumconverter.

**Comment by thomaslevesque** [comment]
To be honest, I'm not very happy with the current solution. `CosmosSerializationOptions` only lets us control a very small subset of JSON.NET's `JsonSerializerSettings`. For instance, I really need to be able to add converters, and the only way to do that now is by reimplementing a complete serializer (which isn't very hard, since I can just copy and modify `CosmosJsonDotNetSerializer`, but it's not ideal).

I understand that you don't want to explicitly depend on JSON.NET, but since there's a package reference, you depend on it anyway.

I think a good approach would be the following:
- implement a default serializer using whatever you want (JSON.NET, System.Text.Json or whatever)
- expose a `CosmosJsonDotNetSerializer` in a separate NuGet package, so users can have more control over the serialization settings

**Comment by bartelink** [comment]
@thomaslevesque 's suggestion will also lead to a thriving STJ implementation sitting alongside it in a matter of days, methinks.
